File: app/src/main/java/edu/scranton/ctleweb/droidtrack/ProjtrackService.java
package edu.scranton.ctleweb.droidtrack;

import java.util.List;

import edu.scranton.ctleweb.droidtrack.projtrack.AllProjectsContent;
import edu.scranton.ctleweb.droidtrack.projtrack.MyProjectsContent;
import retrofit2.Call;
import retrofit2.http.Field;
import retrofit2.http.POST;
import retrofit2.http.Path;

import retrofit2.http.FormUrlEncoded;
import retrofit2.http.Header;
import retrofit2.http.Headers;
import retrofit2.http.GET;

/**
 * Created by sean on 1/9/18.
 */

interface ProjtrackService {
    @FormUrlEncoded
    @POST("get_auth_token/")
    Call<Login> authenticate(@Field("username") String username,
                             @Field("password") String password);
    @Headers("Content-Type: application/json")
    @GET("api/users/{id}/")
    Call<AllProjectsContent.UserItem> getUser(@Path("id") String id,
                                              @Header("authenticate") String token);
    @Headers("Content-Type: application/json")
    @GET("api/projects/{id}/")
    Call<AllProjectsContent.ProjectItem> getProject(@Path("id") String id,
                                                    @Header("authenticate") String token);
    @Headers("Content-Type: application/json")
    @GET("api/clients/{id}/")
    Call<AllProjectsContent.ClientItem> getClient(@Path("id") String id,
                                                  @Header("authenticate") String token);
    @Headers("Content-Type: application/json")
    @GET("api/departments/{id}/")
    Call<AllProjectsContent.DepartmentItem> getDepartment(@Path("id") String id,
                                                          @Header("authenticate") String token);
    @Headers("Content-Type: application/json")
    @GET("api/types/{id}/")
    Call<AllProjectsContent.TypeItem> getType(@Path("id") String id,
                                              @Header("authenticate") String token);
    @Headers("Content-Type: application/json")
    @GET("api/semesters/{id}/")
    Call<AllProjectsContent.SemesterItem> getSemester(@Header("authenticate") String token);

    // List Functions
    @Headers("Content-Type: application/json")
    @GET("api/users/")
    Call<List<AllProjectsContent.UserItem>> getUsers(@Header("authenticate") String token);
    @Headers("Content-Type: application/json")
    @GET("api/all_projects/")
    Call<List<AllProjectsContent.ProjectItem>> getAllProjects(@Header("authenticate") String token);
    @Headers("Content-Type: application/json")
    @GET("api/my_projects/")
    Call<List<MyProjectsContent.ProjectItem>> getMyProjects(@Header("authenticate") String token);
    @Headers("Content-Type: application/json")
    @GET("api/clients/")
    Call<List<AllProjectsContent.ClientItem>> getClients(@Header("authenticate") String token);
    @Headers("Content-Type: application/json")
    @GET("api/departments/")
    Call<List<AllProjectsContent.DepartmentItem>> getDepartments(@Header("authenticate") String token);
    @Headers("Content-Type: application/json")
    @GET("api/types/")
    Call<List<AllProjectsContent.TypeItem>> getTypes(@Header("authenticate") String token);
    @Headers("Content-Type: application/json")
    @GET("api/semesters/")
    Call<List<AllProjectsContent.SemesterItem>> getSemesters(@Header("authenticate") String token);
}

File: app/src/main/java/edu/scranton/ctleweb/droidtrack/ProjtrackService.kt
package edu.scranton.ctleweb.droidtrack

import edu.scranton.ctleweb.droidtrack.projtrack.AllProjectsContent
import edu.scranton.ctleweb.droidtrack.projtrack.Content
import edu.scranton.ctleweb.droidtrack.projtrack.MyProjectsContent
import retrofit2.Call
import retrofit2.http.*

internal interface ProjtrackService {
    @FormUrlEncoded
    @POST("get_auth_token/")
    fun authenticate(@Field("username") username: String?,
                     @Field("password") password: String?): Call<Login>

    @Headers("Content-Type: application/json")
    @GET("api/users/{id}/")
    fun getUser(@Path("id") id: String,
                @Header("authenticate") token: String): Call<Content.UserItem>

    @Headers("Content-Type: application/json")
    @GET("api/projects/{id}/")
    fun getProject(@Path("id") id: String,
                   @Header("authenticate") token: String): Call<AllProjectsContent.ProjectItem>

    @Headers("Content-Type: application/json")
    @GET("api/clients/{id}/")
    fun getClient(@Path("id") id: String,
                  @Header("authenticate") token: String): Call<Content.ClientItem>

    @Headers("Content-Type: application/json")
    @GET("api/departments/{id}/")
    fun getDepartment(@Path("id") id: String,
                      @Header("authenticate") token: String): Call<Content.DepartmentItem>

    @Headers("Content-Type: application/json")
    @GET("api/types/{id}/")
    fun getType(@Path("id") id: String,
                @Header("authenticate") token: String): Call<Content.TypeItem>

    @Headers("Content-Type: application/json")
    @GET("api/semesters/{id}/")
    fun getSemester(@Header("authenticate") token: String): Call<Content.SemesterItem>

    // List Functions
    @Headers("Content-Type: application/json")
    @GET("api/users/")
    fun getUsers(@Header("authenticate") token: String): Call<List<Content.UserItem>>

    @Headers("Content-Type: application/json")
    @GET("api/projects/")
    fun getAllProjects(@Header("authenticate") token: String): Call<List<AllProjectsContent.ProjectItem>>

    @Headers("Content-Type: application/json")
    @GET("api/projects?my_projects=true")
    fun getMyProjects(@Header("authenticate") token: String): Call<List<MyProjectsContent.ProjectItem>>

    @Headers("Content-Type: application/json")
    @GET("api/clients/")
    fun getClients(@Header("authenticate") token: String): Call<List<Content.ClientItem>>

    @Headers("Content-Type: application/json")
    @GET("api/departments/")
    fun getDepartments(@Header("authenticate") token: String): Call<List<Content.DepartmentItem>>

    @Headers("Content-Type: application/json")
    @GET("api/types/")
    fun getTypes(@Header("authenticate") token: String): Call<List<Content.TypeItem>>

    @Headers("Content-Type: application/json")
    @GET("api/semesters/")
    fun getSemesters(@Header("authenticate") token: String): Call<List<Content.SemesterItem>>

    @Headers("Content-Type: application/json")
    @GET("api/current_semester/")
    fun getCurrentSemester(@Header("authenticate") token: String): Call<List<Content.CurrentSemesterItem>>

    @Headers("Content-Type: application/x-www-form-urlencoded")
    @FormUrlEncoded
    @POST("api/projects/")
    fun addProject(@Header("authenticate") token: String,
                   @Field("title") title: String,
                   @Field("description") description: String,
                   @Field("date") date: String,
                   @Field("type") type: Int,
                   @Field("walk_in") walk_in: Boolean,
                   @Field("client") client: Int,
                   @Field("users") users: List<Content.UserItem>,
                   @Field("semester") semester: Int,
                   @Field("hours") hours: String,
                   @Field("completed") completed: Boolean): Call<Void>

    @Headers("Content-Type: application/x-www-form-urlencoded")
    @FormUrlEncoded
    @POST("api/clients/")
    fun addClient(@Header("authenticate") token: String,
                  @Field("first_name") first_name: String,
                  @Field("last_name") last_name: String,
                  @Field("email") email: String,
                  @Field("department") department: Int): Call<Void>
}
